src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +15,135 @@
                     DoCastVictim(SPELL_BANSHEEWAIL);
-                //4 seconds until we should cast this again
-                BansheeWail_Timer = 4000;
-            } else BansheeWail_Timer -= diff;
-
-            //BansheeCurse
-            if (BansheeCurse_Timer <= diff)
-            {
-                if (rand32() % 100 < 75)
+                    events.Repeat(4s);
+                    break;
+                case EVENT_SPELL_BANSHEECURSE:
                     DoCastVictim(SPELL_BANSHEECURSE);
-                //18 seconds until we should cast this again
-                BansheeCurse_Timer = 18000;
-            } else BansheeCurse_Timer -= diff;
-
-            //Silence
-            if (Silence_Timer <= diff)
-            {
-                if (rand32() % 100 < 80)
+                    events.Repeat(18s);
+                    break;
+                case EVENT_SPELL_SILENCE:
                     DoCastVictim(SPELL_SILENCE);
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+                    events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+                    break;
+                case EVENT_SPELL_POSSESS:
+                    if (Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false))
+                    {
+                        DoCast(possessTarget, SPELL_POSSESS, true);
+                        DoCast(possessTarget, SPELL_POSSESSED, true);
+                        DoCastSelf(SPELL_POSSESS_INV, true);
+                        _possessedTargetGuid = possessTarget->GetGUID();
+                        events.ScheduleEvent(EVENT_CHECK_POSSESSED, 0s);
+                    }
+                    else
+                        events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+                    break;
+                case EVENT_INVISIBLE:
+                    // When there's a possessed target
+                    if (Player* possessedTarget = ObjectAccessor::GetPlayer(*me, _possessedTargetGuid))
+                    {
+                        possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                        possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                        me->RemoveAurasDueToSpell(SPELL_POSSESS_INV);
+                        _possessedTargetGuid.Clear();
+                        events.CancelEvent(EVENT_CHECK_POSSESSED);
+                        events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+                    }
+                    break;
+                case EVENT_CHECK_POSSESSED:
+                    if (Player* possessedTarget = ObjectAccessor::GetPlayer(*me, _possessedTargetGuid))
+                    {
+                        if (!possessedTarget->HasAura(SPELL_POSSESSED) || possessedTarget->HealthBelowPct(50))
+                            events.ScheduleEvent(EVENT_INVISIBLE, 0s);
```

Comment:
my earlier point still stands - why is this a separate event? why not just run the "remove auras" logic here?

Reply:
Oh I get you now